### PR TITLE
feat: add support for `InvokeMethod2` to `Parser` and `Weeder`

### PR DESCRIPTION
**Pull request by chanattan**

### CODE REVIEW

**Comment by magnus-madsen** [comment]
@chanattan I see some compiler tests failing. Have a look :)

**Comment by magnus-madsen** [comment]
I see its just some missing cases. Those have to be filled out. 

In Simplifier you should be able to translate InvokeMethod2 to just InvokeMethod.

**Comment by chanattan** [comment]
I filled out the missing cases, can you review them please ?
Also, I didn't change the Simplifier.scala file.

**Comment by magnus-madsen** [comment]
I will let @mlutze take a look, but I think in some of the cases (e.g. redundancy) you should make recursive calls on the object and argument expressions.

File: main/src/ca/uwaterloo/flix/language/ast/ops/TypedAstOps.scala [hunk 1]
```diff
@@ -85,6 +85,7 @@ object TypedAstOps {
     case Expr.TryCatch(exp, rules, _, _, _) => sigSymsOf(exp) ++ rules.flatMap(rule => sigSymsOf(rule.exp))
     case Expr.TryWith(exp, _, rules, _, _, _) => sigSymsOf(exp) ++ rules.flatMap(rule => sigSymsOf(rule.exp))
     case Expr.Do(_, exps, _, _, _) => exps.flatMap(sigSymsOf).toSet
+    case Expr.InvokeMethod2(_, _, exps, _, _, _) => exps.flatMap(sigSymsOf).toSet
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/ast/ops/TypedAstOps.scala:88]: Looks like we are missing `exp` here.

File: main/src/ca/uwaterloo/flix/language/dbg/printer/TypedAstPrinter.scala [hunk 2]
```diff
@@ -76,6 +76,7 @@ object TypedAstPrinter {
     case Expr.TryCatch(exp, rules, _, _, _) => DocAst.Expression.TryCatch(print(exp), rules.map(printCatchRule))
     case Expr.TryWith(exp, _Use, rules, _, _, _) => DocAst.Expression.TryWith(print(exp), _Use.sym, rules.map(printHandlerRule))
     case Expr.Do(op, exps, _, _, _) => DocAst.Expression.Do(op.sym, exps.map(print))
+    case Expr.InvokeMethod2(exp, name, exps, _, _, _) => DocAst.Expression.JavaInvokeMethod2(name, print(exp), exps.map(print)) // TO CHECK
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/dbg/printer/TypedAstPrinter.scala:79]: This seems reasonable to me

File: main/src/ca/uwaterloo/flix/language/dbg/DocAst.scala [hunk 3]
```diff
@@ -255,6 +255,9 @@ object DocAst {
     def Do(sym: Symbol.OpSym, ds: List[Expression]): Expression =
       Keyword("do", App(AsIs(sym.toString), ds))
 
+    def JavaInvokeMethod2(name: Name.Ident, d: Expression, ds: List[Expression]): Expression =
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/dbg/DocAst.scala:258]: This seems reasonable to me.

File: main/src/ca/uwaterloo/flix/language/dbg/printer/TypedAstPrinter.scala (hunk 2 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/dbg/printer/TypedAstPrinter.scala:79]: Except, why do we switch the order of parameters?

File: main/src/ca/uwaterloo/flix/language/phase/EffectVerifier.scala [hunk 4]
```diff
@@ -350,6 +350,8 @@ object EffectVerifier {
       visitExp(exp)
       // TODO ?
       ()
+    case Expr.InvokeMethod2(_, _, _, _, _, _) =>
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/EffectVerifier.scala:353]: Fine for now, but add a `TODO ?` to match the others :)

File: main/src/ca/uwaterloo/flix/language/phase/PatMatch.scala [hunk 5]
```diff
@@ -237,6 +237,7 @@ object PatMatch {
       case Expr.FixpointFilter(_, exp, _, _, _) => visitExp(exp)
       case Expr.FixpointInject(exp, _, _, _, _) => visitExp(exp)
       case Expr.FixpointProject(_, exp, _, _, _) => visitExp(exp)
+      case Expr.InvokeMethod2(_, _, _, _, _, _) => Nil
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/PatMatch.scala:240]: We'll need a recursive visit here.

File: main/src/ca/uwaterloo/flix/language/phase/Redundancy.scala [hunk 6]
```diff
@@ -766,6 +766,9 @@ object Redundancy {
     case Expr.FixpointProject(_, exp, _, _, _) =>
       visitExp(exp, env0, rc)
 
+    case Expr.InvokeMethod2(_, _, _, _, _, _) =>
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/Redundancy.scala:769]: We'll need a recursive visit here.

File: main/src/ca/uwaterloo/flix/language/phase/Regions.scala [hunk 7]
```diff
@@ -279,6 +279,9 @@ object Regions {
     case Expr.FixpointProject(_, exp, tpe, _, loc) =>
       visitExp(exp) ++ checkType(tpe, loc)
 
+    case Expr.InvokeMethod2(_, _, _, _, _, _) =>
+      Nil
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/Regions.scala:283]: Recursive visit

File: main/src/ca/uwaterloo/flix/language/phase/PredDeps.scala [hunk 8]
```diff
@@ -326,6 +326,9 @@ object PredDeps {
     case Expr.FixpointProject(_, exp, _, _, _) =>
       visitExp(exp)
 
+    case Expr.InvokeMethod2(_, _, _, _, _, loc) =>
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/PredDeps.scala:329]: We usually put impossible cases last in the list.

File: main/src/ca/uwaterloo/flix/language/phase/Safety.scala [hunk 9]
```diff
@@ -378,6 +378,9 @@ object Safety {
       case Expr.FixpointProject(_, exp, _, _, _) =>
         visit(exp)
 
+      case Expr.InvokeMethod2(_, _, _, _, _, _) =>
+        Nil
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/Safety.scala:382]: Recursive visit

File: main/src/ca/uwaterloo/flix/language/phase/Stratifier.scala [hunk 10]
```diff
@@ -495,6 +495,9 @@ object Stratifier {
         case e => Expr.FixpointProject(pred, e, tpe, eff, loc)
       }
 
+    case Expr.InvokeMethod2(exp, name, exps, tpe, eff, loc) =>
+      Validation.success((Expr.InvokeMethod2(exp, name, exps, tpe, eff, loc)))
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/Stratifier.scala:499]: Recursive visit

File: main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala [hunk 11]
```diff
@@ -310,6 +310,9 @@ object CodeHinter {
     case Expr.FixpointProject(_, exp, _, _, _) =>
       visitExp(exp)
 
+    case Expr.InvokeMethod2(_, _, _, _, _, _) =>
+      Nil
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala:314]: Recursive visit

**Review by mlutze — COMMENTED** [review]
Missing a few recursive visiting cases.

I also see that we've marked the case as unreachable e.g. in Lowering. But I don't see the case actually getting eliminated in any phase.

**Comment by magnus-madsen** [comment]
@chanattan I will let you work on @mlutze comments and then I will take another look.

Once this is merged in, you can work on Parser2 and start on TypeReduction -- in parallel.

**Comment by chanattan** [comment]
Thank you for your comments. I will do that just later and come back with some questions for the rest.

File: main/src/ca/uwaterloo/flix/language/phase/Regions.scala (hunk 7 above)
Comment (chanattan) [reply main/src/ca/uwaterloo/flix/language/phase/Regions.scala:283]: It seems like exps is not visited in Expr.Apply, I followed the same pattern for InvokeMethod2.

File: main/src/ca/uwaterloo/flix/language/dbg/printer/TypedAstPrinter.scala (hunk 2 above)
Comment (chanattan) [reply main/src/ca/uwaterloo/flix/language/dbg/printer/TypedAstPrinter.scala:79]: It should be changed.

File: main/src/ca/uwaterloo/flix/language/phase/Regions.scala (hunk 7 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Regions.scala:283]: That seems like a bug!!

**Comment by magnus-madsen** [comment]
@chanattan Note: You can mark GitHub comments as resolved as you fix them.

File: main/src/ca/uwaterloo/flix/language/phase/Regions.scala (hunk 7 above)
Comment (chanattan) [reply main/src/ca/uwaterloo/flix/language/phase/Regions.scala]: I see. Then would this be good for both of them ? :
`exps.flatMap(visitExp) ++ visitExp(exp) ++ checkType(tpe, loc)`

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Regions.scala]: Yes

File: main/src/ca/uwaterloo/flix/language/phase/TypeReconstruction.scala [hunk 12]
```diff
@@ -417,6 +417,11 @@ object TypeReconstruction {
       val eff = Type.mkUnion(eff1 :: es.map(_.eff), loc)
       TypedAst.Expr.Do(op, es, tpe, eff, loc)
 
+    case KindedAst.Expr.InvokeMethod2(exp, name, exps, tvar, evar, loc) =>
+      val e = visitExp(exp)
+      val es = exps.map(visitExp(_))
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/TypeReconstruction.scala:422]: visitExp(_) -> visitExp

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/TypeReconstruction.scala:422]: (You can fix this later)

**Comment by magnus-madsen** [comment]
Merged- thanks!

File: main/src/ca/uwaterloo/flix/language/phase/TypeReconstruction.scala (hunk 12 above)
Comment (chanattan) [reply main/src/ca/uwaterloo/flix/language/phase/TypeReconstruction.scala:422]: Thank you. This should be fixed in #7762.

**Pull request merged**